From a1d414a091ed9d1a4e45cdf4caca78d89970a045 Mon Sep 17 00:00:00 2001
From: Jozsef Kiraly <fonix232@gmail.com>
Date: Sun, 27 Sep 2026 05:50:10 +0100
Subject: [PATCH 9/9] arm: dts: qcom: qnap-301w: enable USB (2.0) for rescue
 boots

Enable both USB controllers and their QUSB2 PHYs. U-Boot has no driver
for the IPQ8074 USB3 QMP PHY, so run the ports as USB 2.0 hosts on the
QUSB2 PHY alone; that is enough to load a rescue image from a stick
(fallback level 2). Host only: the DWC3 gadget glue that ARCH_SNAPDRAGON
implies is disabled.
---
 arch/arm/dts/ipq8072-qnap-301w-u-boot.dtsi | 17 +++++++++++++++++
 arch/arm/dts/ipq8072-qnap-301w.dts         | 16 ++++++++++++++++
 configs/qnap_301w_defconfig                | 10 ++++++++++
 3 files changed, 43 insertions(+)

diff --git a/arch/arm/dts/ipq8072-qnap-301w-u-boot.dtsi b/arch/arm/dts/ipq8072-qnap-301w-u-boot.dtsi
index 6b7fe19aac..b752f905c7 100644
--- a/arch/arm/dts/ipq8072-qnap-301w-u-boot.dtsi
+++ b/arch/arm/dts/ipq8072-qnap-301w-u-boot.dtsi
@@ -25,3 +25,20 @@
 		u-boot,mmc-env-partition = "ubootenv";
 	};
 };
+
+/*
+ * U-Boot has no driver for the IPQ8074 USB3 (QMP) PHY: run both ports as
+ * USB 2.0 hosts on the QUSB2 PHY alone, which is enough to load a rescue
+ * image from a stick.
+ */
+&dwc_0 {
+	phys = <&qusb_phy_0>;
+	phy-names = "usb2-phy";
+	maximum-speed = "high-speed";
+};
+
+&dwc_1 {
+	phys = <&qusb_phy_1>;
+	phy-names = "usb2-phy";
+	maximum-speed = "high-speed";
+};
diff --git a/arch/arm/dts/ipq8072-qnap-301w.dts b/arch/arm/dts/ipq8072-qnap-301w.dts
index 38f41cc2df..9a678700a9 100644
--- a/arch/arm/dts/ipq8072-qnap-301w.dts
+++ b/arch/arm/dts/ipq8072-qnap-301w.dts
@@ -100,6 +100,22 @@
 	status = "okay";
 };
 
+&qusb_phy_0 {
+	status = "okay";
+};
+
+&qusb_phy_1 {
+	status = "okay";
+};
+
+&usb_0 {
+	status = "okay";
+};
+
+&usb_1 {
+	status = "okay";
+};
+
 &sdhc_1 {
 	status = "okay";
 	non-removable;
diff --git a/configs/qnap_301w_defconfig b/configs/qnap_301w_defconfig
index 14df0a23c3..619e72832f 100644
--- a/configs/qnap_301w_defconfig
+++ b/configs/qnap_301w_defconfig
@@ -37,6 +37,7 @@ CONFIG_CMD_GPIO=y
 CONFIG_CMD_MMC=y
 CONFIG_CMD_PART=y
 CONFIG_CMD_READ=y
+CONFIG_CMD_USB=y
 CONFIG_CMD_EXT4=y
 CONFIG_CMD_FAT=y
 CONFIG_CMD_FS_GENERIC=y
@@ -61,6 +62,8 @@ CONFIG_MISC=y
 CONFIG_MMC_SDHCI=y
 CONFIG_MMC_SDHCI_ADMA=y
 CONFIG_MMC_SDHCI_MSM=y
+CONFIG_PHY=y
+CONFIG_PHY_QCOM_QUSB2=y
 CONFIG_PINCTRL=y
 CONFIG_PINCONF=y
 CONFIG_PINCTRL_QCOM_IPQ8074=y
@@ -68,3 +71,10 @@ CONFIG_DEBUG_UART_ANNOUNCE=y
 CONFIG_DEBUG_UART_SKIP_INIT=y
 CONFIG_MSM_SERIAL=y
 CONFIG_SOC_QCOM=y
+CONFIG_USB=y
+# CONFIG_DM_USB_GADGET is not set
+CONFIG_USB_XHCI_HCD=y
+CONFIG_USB_XHCI_DWC3=y
+CONFIG_USB_DWC3=y
+CONFIG_USB_DWC3_GENERIC=y
+CONFIG_USB_STORAGE=y
